cmd/juju/application/binding.go

```diff
@@ -0,0 +1,133 @@
+	"github.com/juju/errors"
+)
+
+const parseBindErrorPrefix = "--bind must be in the form '[<default-space>] [<endpoint-name>=<space> ...]'. "
+
+// parseBindExpr parses the --bind option and returns back a map where keys
+// are endpoint names and values are space names. Valid forms are:
+// * relation-name=space-name
+// * extra-binding-name=space-name
+// * space-name (equivalent to binding all endpoints to the same space, i.e. application-default)
+// * The above in a space separated list to specify multiple bindings,
+//   e.g. "rel1=space1 ext1=space2 space3"
+func parseBindExpr(expr string, knownSpaces []string) (map[string]string, error) {
+	if expr == "" {
+		return nil, nil
+	}
+
+	knownSpaceMap := make(map[string]struct{})
+	for _, spaceName := range knownSpaces {
+		knownSpaceMap[spaceName] = struct{}{}
+	}
+
+	parsedBindings := make(map[string]string)
+	for _, s := range strings.Fields(expr) {
+		v := strings.Split(s, "=")
+		var endpoint, spaceName string
+		switch len(v) {
+		case 1:
+			endpoint = ""
+			spaceName = v[0]
+		case 2:
+			if v[0] == "" {
+				return nil, errors.New(parseBindErrorPrefix + "Found = without endpoint name. Use a lone space name to set the default.")
+			}
+			endpoint = v[0]
+			spaceName = v[1]
+		default:
+			return nil, errors.New(parseBindErrorPrefix + "Found multiple = in binding. Did you forget to space-separate the binding list?")
+		}
+
+		// This is a temporary hack to allow us to bind endpoints to the
+		// default spaceName which is currently represented as a blank string.
+		// It will be removed when the work for properly naming the
+		// default spaceName lands.
+		spaceName = strings.Trim(spaceName, `"`)
+
+		if _, exists := knownSpaceMap[spaceName]; !exists {
+			return nil, errors.NotFoundf("Space with name %q", spaceName)
+		}
+
+		parsedBindings[endpoint] = spaceName
+	}
+	return parsedBindings, nil
+}
+
+// mergeBindings is invoked when upgrading a charm to merge the existing set of
+// endpoint to space assignments for a deployed application and the user-defined
+// bindings passed to the upgrade-charm command.
+func mergeBindings(newCharmEndpoints set.Strings, oldEndpointsMap, userBindings map[string]string, oldDefaultSpace string) (map[string]string, []string) {
+	var changelog []string
+	if oldEndpointsMap == nil {
+		oldEndpointsMap = make(map[string]string)
+	}
+	if userBindings == nil {
+		userBindings = make(map[string]string)
+	}
+
+	newDefaultSpace, newDefaultBindingDefined := userBindings[""]
+	changeDefaultSpace := newDefaultBindingDefined && oldDefaultSpace != newDefaultSpace
+	mergedBindings := make(map[string]string)
+	unmodifiedEndpoints := make(map[string][]string)
+	for _, epName := range newCharmEndpoints.SortedValues() {
+		newSpaceAssignment, newBindingDefined := userBindings[epName]
+
+		// If this is a new endpoint and it is not specified, assign it to the default space
+		oldSpaceAssignment, isOldEndpoint := oldEndpointsMap[epName]
+		if !isOldEndpoint {
```

Comment:
These `if`s might read better as a `switch` without the `continue`s

Reply:
Normally I would agree. However in this case I prefer this format as it allows for more verbose comments about the conditions that we are testing. I think they would be more noisy inside a case block. Thoughts?